docs/proposals/08-shoot-additional-container-runtimes.md

```diff
@@ -0,0 +1,159 @@
+    2. os-coreos - 
+        1. Add a controlplane webhook to manipulate the OperatingSystemConfig and add the following flags to the kubelet execution command:
+           1. --container-runtime=remote
+           2. --container-runtime-endpoint=unix:///var/run/docker/libcontainerd/docker-containerd.socket
+        2. Create a containerd configuration file: /etc/containerd/config.toml based on the default configuration of containerd. 
+        3. Add a generator that will override the default containerd.service unit. It will create a new unit in: /etc/systemd/system/containerd.service unit that will start the containerd service with a configuration file path: /etc/containerd/config.toml.
+        
+    3. os-suse-jeos - __TBD__ 
+    
+3. Docker-monitor daemon and rotate log should be replaced with equivalent conatinerd services. __TBD__ 
+4. In order to install additional runtimes on the worker machine we must provide the installation scripts to the machine file system and run them.
+   The installation script should download and install the runtime binaries in each OS.
+   Instructions for Gardener to copy files to the machine filesystem and execute them is done via the cloud config secret.
+   
+   1. Each OS extension should implement a generator that will copy a generic packages installation script file.
+        ```yaml
+          #cloud-config
+          write_files:
+            path: '/var/lib/packages/install.sh'
+            ...
+      ```
+      
+      The installation file will be executed with different parameters defining the relevant package that should be installed:
+        1. Package name
+        2. Package path url
+        3.  __TBD__ 
+        
+   2. Each Extension controller should reconcile the OperatingSystemConfig (Should be enhanced to contain additional runtime data __TBD__):
+        1. Create a service unit to run the runtime binaries package installation script with the relevant parameters (e.g name of runtime container. package url).
+           ```
+           [Unit]
+           Description=Download and install additional runtimes binaries
+           ...
+       
+           [Service]
+           ExecStart=/var/lib/packages/install.sh <Parameters>
+           ...
+           ```
+        2. Copy a container runtime configuration script file and create a unit to run it. For example:
+           ```yaml
+           #cloud-config
+           write_files:
+           path: '/var/lib/kata-containers/configure.sh'
+           ...
+           [Unit]
+           Description=Configure 
+           ...
+      
+           [Service]
+           ExecStart=/var/lib/kata-containers/configure.sh
+           ```
+           
+           The installation script will be responsible for configuring the config.toml plugins at /etc/containerd/config.toml to contain the relevant runtime plugins.
+                1. kata-containers -  __TBD__
+                2. gvisor - __TBD__
+  
```

Comment:
It is highly recommended to avoid installation scripts. Especially dynamic ones and especially such scripts downloading stuff from somewhere. This explicitly contradicts the way we follow to support closed or on-prem environments. There is a defined way how to deal with container images. And all other components (like the CNI components) install their stuff by copying it by running containers. This should be kept here. I see the problem, that running a container requires a working container engine.
But we should have a closer look at the possibilities here and invest some research:
- using the engine coming with the OS image to run installation containers (before the kubelet is starting)
- or even may be:  starting the kubelet with a standard environment available everywhere, which then starts a daemon set reconfiguring the kubelet.

Reply:
This installation isn't needed to start the kubelet itself. Nor do we need to reconfigure kubelet to make it possible to use these additional runtimes. Kubelet will be run using `containerd`, which is installed as part of the `Docker Engine` and is already on the base OS images. Making Kubernetes run with `containerd` is the first part of the proposal, and doesn't require installing anything on the worker nodes.

This point behind this installation is to make it possible for the container abstraction (`containerd` in our case) to run containers using a different low-level implementation (gVisor or Kata Containers) as opposed to the default docker cgroups-based low-level implementation. It's simply not possible to run this inside a container. The engine setting up the container (`containerd`) needs to have direct access to the binaries, which are required to be installed on the host file-system where the `containerd` daemon itself is running.  

We'll research if it's possible to use OCI images as a transport mechanism - pull an image and then extract it into the host filesystem so that the binaries it contains can be used. 

Another alternative would be to package the relevant binaries (as tarballs) as part of the gardener extension and download the binaries from there

Both these alternatives would make it possible to support closed or on-prem environments .

Reply:
We will get the installation from OCI Image provided in a deamonset